core/src/main/java/org/elasticsearch/index/store/Store.java

```diff
@@ -1295,14 +1308,20 @@ private void readAndCompareChecksum() throws IOException {
 
         @Override
         public void writeBytes(byte[] b, int offset, int length) throws IOException {
-            if (writtenBytes + length > checksumPosition && actualChecksum == null) {
-                assert writtenBytes <= checksumPosition;
-                final int bytesToWrite = (int) (checksumPosition - writtenBytes);
-                out.writeBytes(b, offset, bytesToWrite);
-                readAndCompareChecksum();
-                offset += bytesToWrite;
-                length -= bytesToWrite;
-                writtenBytes += bytesToWrite;
+            if (writtenBytes + length > checksumPosition) {
+                if (actualChecksum == null) {
+                    assert writtenBytes <= checksumPosition;
+                    final int bytesToWrite = (int) (checksumPosition - writtenBytes);
+                    out.writeBytes(b, offset, bytesToWrite);
+                    readAndCompareChecksum();
+                    offset += bytesToWrite;
+                    length -= bytesToWrite;
+                    writtenBytes += bytesToWrite;
+                }
+                long checksumIndex = writtenBytes - checksumPosition;
+                for (int i = 0;checksumIndex < footerChecksum.length && i<length; checksumIndex++, i++) { // checksum is written
+                    footerChecksum[(int)checksumIndex] = b[offset+i];
+                }
```

Comment:
What do you think of simplifying the last few lines with the following?

    int checksumIndex = Math.toIntExact(writtenBytes - checksumPosition);
    System.arraycopy(b, offset, footerChecksum, checksumIndex, length);